Add support for single go version

egieseke opened the pull request:
## Summary

go.mod is now used to specify the go version. 

## Test Plan

Ran `make clean install`, `make test`, `make integration`, and `mule -f mule.yaml build-linux-arm64`

Requires the latest version of Mule.

btoll commented on scripts/get_golang_version.sh:
```diff
@@ -0,0 +1,2 @@
+#!/usr/bin/env bash
+echo $(cat ./go.mod | grep 'go ' | head -n 1 | cut -d' ' -f2)
```
Here are some simpler alternatives:
```
sed -n 's/^go\s\([.0-9]*\)$/\1/p' go.mod

awk '/^go /{ print $0 }' go.mod | cut -d\  -f2
```
There's probably a way to do it with `awk` in one go with group pattern matching where it's not having to use another process like `cut`.

Also, I'm pretty sure that `shellcheck` would have flagged your command above.  Have you installed it and integrated it with your editor?

egieseke replied:
Removed the use of `cat` to satisfy `shellcheck`.

egieseke replied:
Updated to use awk which is cleaner.  Checked with shellcheck.

btoll replied:
That looks great.

A couple small things.  It's not necessary to capture the output of `awk` in a variable and `echo` that.  Just have the `awk` command output to its caller.

Also, that regex should be tightened, it's a bit "hole-y" right now.  Something like this would make sure that only the desired line is selected:
```
awk '/^go\s[0-9.]+/{print $2}' go.mod
```

egieseke replied:
Updated awk regex to be more specific.

btoll replied:
The `echo` is unnecessary.  Just have the `awk` command.

Please integrate your editor with `shellcheck` or install a githook.  It will save this kind of back and forth.

While it works in most cases, your regex is overly-complex and does fails some tests.  The one I pasted in above is simple and just uses one character class, matches any number of {number}. pairs, and matches any whitespace (actually, I modified it a bit to match any number of whitespace before the number sequence, we don't really care after it).

```
awk '/^go\s*[0-9.]+/{print $2}' go.mod
```

egieseke replied:
Removed echo.  What tests does the current awk command not support?

btoll replied:
- It would fail if the tooling changed to place any text after the version string.  I think it would be better to match what we need rather than the whole line, which is what your regex is doing with the `$` anchor.

- It wouldn't catch versions that have four positions, i.e., `1.12.34.56`.

egieseke replied:
This can be updated when and if the tooling changes or go starts supporting 4 levels of version numbers.  Currently, the go version supports major, minor, and optional patch number.  What you are suggesting is more "hole-y".

btoll replied:
Ok, sounds good.

egieseke commented:
> Changes here looks good, but there is one thing missing -
> you'll need to update the .travis file and remove the following section:
> 
> ```
> go:
>   - "1.12"
> go_import_path: github.com/algorand/go-algorand
> language: go
> ```
> 
> than, you'll need to add the installation portion to `scripts/travis/configure_dev.sh`

@tsachiherman, I removed those lines from the .travis.yaml file and updated configure_dev.sh to use the correct go version.

tsachiherman commented:
> > Changes here looks good, but there is one thing missing -
> > you'll need to update the .travis file and remove the following section:
> > ```
> > go:
> >   - "1.12"
> > go_import_path: github.com/algorand/go-algorand
> > language: go
> > ```
> > 
> > 
> > than, you'll need to add the installation portion to `scripts/travis/configure_dev.sh`
> 
> @tsachiherman, I removed those lines from the .travis.yaml file and updated configure_dev.sh to use the correct go version.

I don't see where you're installing go on amd64 platforms. ( which are the *only* platforms that runs on travis directly ).
The arm and arm64 changes are used on the build host instances only.

winder commented:
The go version specified in the README could also be updated, it currently specifies as 1.12.x.

Also it looks like you have a flat file specifying `1.14`, but the go.mod says `1.12.17`. Which is it?

tsachiherman commented on scripts/travis/configure_dev.sh:
```diff
@@ -11,9 +11,7 @@ function installGo() {
-        echo "Correct Go version cannot be found; downloading go $GO_VERSION for ${OS_ARCH}..."
-        # go is not installed ?
-        # e.g. https://dl.google.com/go/go1.13.5.linux-arm64.tar.gz
+        echo "Correct Go version not be found; downloading go $GO_VERSION for ${OS_ARCH}..."
```
I'm alittle bit puzzled here.. usually we fix typos, not introduce these intentinally.

egieseke replied:
Not intentional, fixed.

egieseke commented:
> The go version specified in the README could also be updated, it currently specifies as 1.12.x.
> 
> Also it looks like you have a flat file specifying `1.14`, but the go.mod says `1.12.17`. Which is it?

Updated readme to reference go.mod file with a link to the file.  Removed unnecessary go_version.txt file.  The go.mod file is now the authoritative source for go version.

tsachiherman commented on scripts/get_installed_golang_version.sh:
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+if go version > /dev/null 2>&1 ; then
```
it would be nice if you could extract the go version information in a single call rather than making two calls.
( but it looks like it would work correct either way )

egieseke replied:
@tsachiherman Updated to make a single call to go version.

tsachiherman commented on scripts/travis/configure_dev.sh:
```diff
@@ -3,60 +3,56 @@
+    if [[ "${INSTALLED_GO_VERSION}" != "${GO_VERSION}" ]]; then
+        echo "Installing go version ${GO_VERSION} to replace ${INSTALLED_GO_VERSION}"
+        if [[ "${ARCH}" = "amd64" ]]; then
+            eval "$(gimme ${GO_VERSION})"
```
Using gimme here is ok, however, it creates a different execution path for amd64 vs non-amd64.
If you could have "taken" the gimme script and use it in all the cases, it would create a more uniform script.

egieseke replied:
@tsachiherman Updated go version management for Travis builds to use gimme for all platforms.

tsachiherman commented on scripts/travis/configure_dev.sh:
```diff
@@ -10,12 +10,11 @@ function installGo() {
-            sudo apt-get update -y
-            sudo apt-get -y install gimme
+        if [[ "${OS_ARCH}" == "linux-arm64" || "${OS_ARCH}" == "linux-armv6l" ]]; then
+            sudo apt install ruby ruby-dev
```
don't you think that it's going a little bit too far ?
i.e. you're installing ruby, so that you can install travis, so that you could use gimme so that you could install go

maybe you should just download and install go as before ?
I'm concerned that you'll make the build process too long. Right now, the linux-armv6l already takes 6+ hours. Adding these would clearly make it longer..

egieseke replied:
@tsachiherman simplified downloading gimme.

tsachiherman commented:
Awesome. Could you try running it also against the armv6 builder on the
buildhost to ensure that it works?
( i.e. it should.. )

On Thu, Jun 25, 2020, 00:07 egieseke <notifications@github.com> wrote:

> *@egieseke* commented on this pull request.
> ------------------------------
>
> In scripts/travis/configure_dev.sh
> <https://github.com/algorand/go-algorand/pull/1185#discussion_r445295288>:
>
> > @@ -10,12 +10,11 @@ function installGo() {
>      echo "Ensure Go version ${GO_VERSION} for platform ${OS_ARCH}"
>      if [[ "${INSTALLED_GO_VERSION}" != "${GO_VERSION}" ]]; then
>          echo "Installing go version ${GO_VERSION} to replace ${INSTALLED_GO_VERSION}"
> -        if [[ "${OS_ARCH}" = "linux-arm64" || "${OS_ARCH}" = "linux-armv6l" ]]; then
> -            sudo apt-get update -y
> -            sudo apt-get -y install gimme
> +        if [[ "${OS_ARCH}" == "linux-arm64" || "${OS_ARCH}" == "linux-armv6l" ]]; then
> +            sudo apt install ruby ruby-dev
>
> @tsachiherman <https://github.com/tsachiherman> simplified downloading
> gimme.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/algorand/go-algorand/pull/1185#discussion_r445295288>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF2OOH4SPSFVYDOO5S5YXR3RYLERVANCNFSM4OCFG3OQ>
> .
>

egieseke commented:
> Awesome. Could you try running it also against the armv6 builder on the buildhost to ensure that it works? ( i.e. it should.. )
> […](#)
> On Thu, Jun 25, 2020, 00:07 egieseke ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In scripts/travis/configure_dev.sh <[#1185 (comment)](https://github.com/algorand/go-algorand/pull/1185#discussion_r445295288)>: > @@ -10,12 +10,11 @@ function installGo() { echo "Ensure Go version ${GO_VERSION} for platform ${OS_ARCH}" if [[ "${INSTALLED_GO_VERSION}" != "${GO_VERSION}" ]]; then echo "Installing go version ${GO_VERSION} to replace ${INSTALLED_GO_VERSION}" - if [[ "${OS_ARCH}" = "linux-arm64" || "${OS_ARCH}" = "linux-armv6l" ]]; then - sudo apt-get update -y - sudo apt-get -y install gimme + if [[ "${OS_ARCH}" == "linux-arm64" || "${OS_ARCH}" == "linux-armv6l" ]]; then + sudo apt install ruby ruby-dev @tsachiherman <https://github.com/tsachiherman> simplified downloading gimme. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#1185 (comment)](https://github.com/algorand/go-algorand/pull/1185#discussion_r445295288)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AF2OOH4SPSFVYDOO5S5YXR3RYLERVANCNFSM4OCFG3OQ> .

@tsachiherman The build is still running, but gimme seems to work ok for armv6:  https://travis-ci.com/github/algorand/go-algorand/builds/173008734, where the go version is updated to `go version go1.12.17 linux/arm`.

tsachiherman commented:
> > Awesome. Could you try running it also against the armv6 builder on the buildhost to ensure that it works? ( i.e. it should.. )
> > […](#)
> > On Thu, Jun 25, 2020, 00:07 egieseke _**@**_._**> wrote: _**@**_.**_* commented on this pull request. ------------------------------ In scripts/travis/configure_dev.sh <[#1185 (comment)](https://github.com/algorand/go-algorand/pull/1185#discussion_r445295288)>: > @@ -10,12 +10,11 @@ function installGo() { echo "Ensure Go version ${GO_VERSION} for platform ${OS_ARCH}" if [[ "${INSTALLED_GO_VERSION}" != "${GO_VERSION}" ]]; then echo "Installing go version ${GO_VERSION} to replace ${INSTALLED_GO_VERSION}" - if [[ "${OS_ARCH}" = "linux-arm64" || "${OS_ARCH}" = "linux-armv6l" ]]; then - sudo apt-get update -y - sudo apt-get -y install gimme + if [[ "${OS_ARCH}" == "linux-arm64" || "${OS_ARCH}" == "linux-armv6l" ]]; then + sudo apt install ruby ruby-dev @tsachiherman https://github.com/tsachiherman simplified downloading gimme. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#1185 (comment)](https://github.com/algorand/go-algorand/pull/1185#discussion_r445295288)>, or unsubscribe https://github.com/notifications/unsubscribe-auth/AF2OOH4SPSFVYDOO5S5YXR3RYLERVANCNFSM4OCFG3OQ .
> 
> @tsachiherman The build is still running, but gimme seems to work ok for armv6: https://travis-ci.com/github/algorand/go-algorand/builds/173008734, where the go version is updated to `go version go1.12.17 linux/arm`.

To play it safe, just wait until the armv6 build is over and monitor the generated log file to ensure it didn't fail due to some other issue. If it passes, I think that this PR is good to go.

egieseke commented:
> > > Awesome. Could you try running it also against the armv6 builder on the buildhost to ensure that it works? ( i.e. it should.. )
> > > […](#)
> > > On Thu, Jun 25, 2020, 00:07 egieseke _**@**_._**> wrote: _**@**_.**_* commented on this pull request. ------------------------------ In scripts/travis/configure_dev.sh <[#1185 (comment)](https://github.com/algorand/go-algorand/pull/1185#discussion_r445295288)>: > @@ -10,12 +10,11 @@ function installGo() { echo "Ensure Go version ${GO_VERSION} for platform ${OS_ARCH}" if [[ "${INSTALLED_GO_VERSION}" != "${GO_VERSION}" ]]; then echo "Installing go version ${GO_VERSION} to replace ${INSTALLED_GO_VERSION}" - if [[ "${OS_ARCH}" = "linux-arm64" || "${OS_ARCH}" = "linux-armv6l" ]]; then - sudo apt-get update -y - sudo apt-get -y install gimme + if [[ "${OS_ARCH}" == "linux-arm64" || "${OS_ARCH}" == "linux-armv6l" ]]; then + sudo apt install ruby ruby-dev @tsachiherman https://github.com/tsachiherman simplified downloading gimme. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#1185 (comment)](https://github.com/algorand/go-algorand/pull/1185#discussion_r445295288)>, or unsubscribe https://github.com/notifications/unsubscribe-auth/AF2OOH4SPSFVYDOO5S5YXR3RYLERVANCNFSM4OCFG3OQ .
> > 
> > 
> > @tsachiherman The build is still running, but gimme seems to work ok for armv6: https://travis-ci.com/github/algorand/go-algorand/builds/173008734, where the go version is updated to `go version go1.12.17 linux/arm`.
> 
> To play it safe, just wait until the armv6 build is over and monitor the generated log file to ensure it didn't fail due to some other issue. If it passes, I think that this PR is good to go.

@tsachiherman The arm build completed successfully: https://travis-ci.com/github/algorand/go-algorand/builds/173008734

bricerisingalgorand commented:
https://travis-ci.com/github/algorand/go-algorand/builds/173008734#L410

@tsachiherman We ran a job doing only the arm build with these changes and it worked 👍

tsachiherman commented:
> https://travis-ci.com/github/algorand/go-algorand/builds/173008734#L410
> 
> @tsachiherman We ran a job doing only the arm build with these changes and it worked 👍

that's great.. did you review the log ? The last few lines there are : 
```
Still waiting for build to complete 104m:3s...
Build is taking too long. Travis is going to timeout this build, so we'll tell travis that we're done for now.
Once this build is complete, you could get a complete log by typing:
./scripts/travis/external_build_printlog.sh "s3://algorand-build-pull-requests/linux-arm/46900001-log" "s3://algorand-build-pull-requests/linux-arm/46900001-completed.json" --no-sign-request
```

did you try to follow that ?

egieseke commented:
> > https://travis-ci.com/github/algorand/go-algorand/builds/173008734#L410
> > @tsachiherman We ran a job doing only the arm build with these changes and it worked 👍
> 
> that's great.. did you review the log ? The last few lines there are :
> 
> ```
> Still waiting for build to complete 104m:3s...
> Build is taking too long. Travis is going to timeout this build, so we'll tell travis that we're done for now.
> Once this build is complete, you could get a complete log by typing:
> ./scripts/travis/external_build_printlog.sh "s3://algorand-build-pull-requests/linux-arm/46900001-log" "s3://algorand-build-pull-requests/linux-arm/46900001-completed.json" --no-sign-request
> ```
> 
> did you try to follow that ?
> 
> as a hint, it usually takes ~5-6 hours for it to complete. If you figured out that it's working in less time, you probably didn't wait for it to be "done".

@tsachiherman Upgrading on arm32 is problematic due to the installed version of TLS which is less than 1.2 and no longer supported by golang.org.  I created a [separate ticket](https://github.com/algorand/go-algorand-internal/issues/620) to update the build VM to upgrade the OpenSSL version.  For now, disabling changing the go version for arm32 so that this PR can be committed per Tsachi's recommendation.

tsachiherman commented on scripts/travis/build.sh:
```diff
@@ -34,6 +35,12 @@ SCRIPTPATH="$( cd "$(dirname "$0")" ; pwd -P )"
 OS=$("${SCRIPTPATH}/../ostype.sh")
 ARCH=$("${SCRIPTPATH}/../archtype.sh")
 
+if [[ "${ARCH}" != "arm" ]]; then
+    curl -sL -o ~/gimme https://raw.githubusercontent.com/travis-ci/gimme/master/gimme
+    chmod +x ~/gimme
+    eval $(~/gimme $("${SCRIPTPATH}/../get_golang_version.sh"))
+fi
+
```
could you move this code into the `scripts/travis/configure_dev.sh` so that it would get attempted multiple times ?
( travis sometimes has some temporary issues, which were proved to get solved by trying few times )

egieseke replied:
@tsachiherman the gimme script will not configure the environment correctly from a subshell.

tsachiherman replied:
so.. use the `ln` trick that I used.. it will make it working correctly, subshell or not..

egieseke replied:
Using symbolic links only works when go is already installed, and the location of the links vary based on the platform.  One advantage of using the environment context is that we can now test multiple versions of go in parallel.

tsachiherman commented on scripts/travis/test.sh:
```diff
@@ -1,11 +1,18 @@
 OS=$("${SCRIPTPATH}/../ostype.sh")
 ARCH=$("${SCRIPTPATH}/../archtype.sh")
 
+if [[ "${ARCH}" != "arm" ]]; then
+    curl -sL -o ~/gimme https://raw.githubusercontent.com/travis-ci/gimme/master/gimme
+    chmod +x ~/gimme
+    eval $(~/gimme $("${SCRIPTPATH}/../get_golang_version.sh"))
+fi
+
```
The `travis/test.sh` is called only after `travis/build.sh` which already have this installed.

egieseke replied:
Required due to environment scope of gimme.  Gimme reuses go install if it exists already.

tsachiherman commented on scripts/travis/integration_test.sh:
```diff
@@ -9,7 +9,16 @@
 SCRIPTPATH="$( cd "$(dirname "$0")" ; pwd -P )"
+ARCH=$("${SCRIPTPATH}/../archtype.sh")
+
+if [[ "${ARCH}" != "arm" ]]; then
+    curl -sL -o ~/gimme https://raw.githubusercontent.com/travis-ci/gimme/master/gimme
+    chmod +x ~/gimme
+    eval $(~/gimme $("${SCRIPTPATH}/../get_golang_version.sh"))
+fi
+
```
The `travis/integration_test.sh` is called only after `travis/build.sh` which already have this installed.

egieseke replied:
> The `travis/integration_test.sh` is called only after `travis/build.sh` which already have this installed.


@tsachiherman The `travis/integration_test.sh` script is invoked directly from the .travis.yaml file.

tsachiherman replied:
sorry, my bad. But given that this script is already calling `travis/build.sh` few lines below, why do we need it ?

egieseke replied:
@tsachiherman 
Without, results in an error on the external ARM64 integration test.  
`++ go env GOPATH
./scripts/travis/integration_test.sh: line 32: go: command not found`

tsachiherman replied:
@egieseke , you replaced the install that we previously had with your custom installer, and then after your custom installer doesn't install it correctly, you figure out that you need to install it over and over..
The reason I created the symbolic link when installing go was to work around setting the PATH, which cannot be done from a subshell. That's also the case here where you've seen the error coming from
```bash
GOPATHBIN=$(go env GOPATH)/bin
```
because it couldn't find `go` in any of the `PATH` locations.

egieseke replied:
@tsachiherman Because we have multiple entry points for building and testing with Travis, it must be invoked once at the top level to set up the go environment. Note that the gimme script is also used by Travis to configure the go environment.

egieseke replied:
@tsachiherman Removed call to gimme from integration_test.sh.

tsachiherman commented on config/consensus.go:
```diff
@@ -496,7 +496,7 @@ func initConsensusProtocols() {
 		MaxBalLookback: 320,
 
-		MaxTxGroupSize:               1,
+		MaxTxGroupSize: 1,
```
could you undo all unintended changes ? they won't contribute anything for supporting a single go version.
( if you really think we should make these changes, move these to a separate PR )

egieseke replied:
Reverted unintended formatting.

egieseke commented:
> > > https://travis-ci.com/github/algorand/go-algorand/builds/173008734#L410
> > > @tsachiherman We ran a job doing only the arm build with these changes and it worked 👍
> > 
> > 
> > that's great.. did you review the log ? The last few lines there are :
> > ```
> > Still waiting for build to complete 104m:3s...
> > Build is taking too long. Travis is going to timeout this build, so we'll tell travis that we're done for now.
> > Once this build is complete, you could get a complete log by typing:
> > ./scripts/travis/external_build_printlog.sh "s3://algorand-build-pull-requests/linux-arm/46900001-log" "s3://algorand-build-pull-requests/linux-arm/46900001-completed.json" --no-sign-request
> > ```
> > 
> > 
> > did you try to follow that ?
> > as a hint, it usually takes ~5-6 hours for it to complete. If you figured out that it's working in less time, you probably didn't wait for it to be "done".
> 
> @tsachiherman Upgrading on arm32 is problematic due to the installed version of TLS which is less than 1.2 and no longer supported by golang.org. I created a [separate ticket](https://github.com/algorand/go-algorand-internal/issues/620) to update the build VM to upgrade the OpenSSL version. For now, disabling changing the go version for arm32 so that this PR can be committed per Tsachi's recommendation.

@tsachiherman Updated arm32 AMI to use OpenSSL 1.1.1g with TLS 1.3.  After this change, the arm32 build completed successfully with go 1.12.17.
